### DISCUSSION TITLE
How does the actual Next "backend" work on Vercel?

### DISCUSSION THREAD

**Discussion by kieri097** [vercel/next.js#21836] [Help] [answered]
Hi,

Going through the Vercel page i found this:

> Next.js bundles all API Routes and server-rendered pages into shared Serverless Functions , separating API routes and server-rendered pages when deployed on Vercel. Further Serverless Functions are created only if they do not fit within the 50MB limit

Just to understand the logic here (let's ignore the bundling on Vercel's side for now):

Is it correct that for every SSR page à la [id].js, one serverless function will be in charge of generating the rendered page? And for every SSRender, the lambda would be triggered? 

How does that work for ISR with a revaluate? Let's assume I have 5 pages of which I pre-build 2, 2 are eventually built, and the last one is never triggered. Would that spin up 5 serverless functions and call the 2 active pages on every revaluate until the other 2 pages are eventually opened, in which case all 4 "activated" lambdas are called? What about  the "dormant" page's lambda? I assume that gets spun up as well but is never triggered until it is activated, or does it still run according to the revaluate schedule? 

Thanks!

**Comment by cobrexus** [comment, marked as answer]
Each time a `getServerSideProps` function is run, it is a serverless function being spun up.

Each time a `getStaticProps` with `revalidate` set is run, it is also a serverless function being spun up.

Vercel uses AWS Lambda for serverless functions.

The way a lambda or serverless function works is that it initiates the Node.js environment, runs the code, and then closes it. Functions are not kept running.

**Reply by kieri097** [reply]
Thanks for this!

**Reply by utibeabasi6** [reply]
hey @aryanbeezadhur does vercel package the entire app source code into a lambda? im trying to understand how it would import functions defined in external files if it is creating a lambda for every page

**Reply by cobrexus** [reply]
@utibeabasi6 when I wrote my answer, Next.js 13 and the App router wasn't out yet, and I'm not sure how it differs from the Pages router as I decided to stick with the Pages router for now.

**Reply by utibeabasi6** [reply]
alright thanks
